_posts/2022-05-18-My-OTP-25-highlights.md

```diff
@@ -0,0 +1,751 @@
+
+Some applications in OTP like SSL/TLS and SSH need cryptography to
+work. That is provided by the OTP application crypto, which interfaces
+Erlang to an external cryptolib in C using NIFs. The main example of
+such an external cryptolib is [OpenSSL](https://www.openssl.org).
+
+The OpenSSL cryptolib exists in many versions. OTP/crypto supports
+0.9.8c and later, although only 1.1.1 is still maintained by OpenSSL.
+
+OpenSSL has released its version 3.0 series, which is their future
+platform totally re-built with a new API. The APIs of previous
+versions (1.1.1 and older) are partly deprecated, although still
+available in 3.0. The support of 1.1.1 will also end in a future.
+
+Since it is vital to get security patches in the cryptolib, and in a
+future only the 3.0 API might be available, OTP/crypto now from
+OTP-25.0 interfaces OpenSSL 3.0 using the new 3.0 API. A few functions
+from old APIs are still used, but they will be replaced as soon as
+possible.
+
+You as a user will hopefully not notice any difference: if you have
+OpenSSL 1.1.1 (or older - not recommended) and build OTP, that one will
+be used as previously. If you have any OpenSSL 3.0 version installed,
+that one will be used without need of doing anything special except
+for normal handling of dynamic loading paths in the OS.
+
+Although OTP/crypto's interface to the new API is heavily tested on many
+architectures for a long time, we do not yet recommend it for usage in
+critical applications. In such ones it is still safer to build crypto with
+the latest 1.1.1 release.
+
+# CA-certificates can be fetched from the OS standard place
+
+With the new functions `public_key:cacerts_load/0,1` and `public_key:cacerts_get/0` the CA certificates can be fetched from the standard place of the OS (or from a file). 
+
+They will then be cached in decoded form by use of `persistent_term` which makes them available in an efficient way for the `ssl` and `httpc` modules. The intention with this is to make it unnecessary to depend on for example `certifi` in many packages.
+
+On Windows and MacOSx the certificate store is not an ordinary file so the information is fetched via an API using a NIF (Windows) or with an external program (MacOSx). 
+
+Example with `ssl`
+```erlang
+public_key:cacerts_load(), % a noop if certs are already loaded
+%% makes the certificates available without copying
+CaCerts = public_key:cacerts_get(), 
+% use the certificates when establishing a connection
+{ok,Socket} = ssl:connect("erlang.org",80,[{cacerts,CaCerts}, {verify,verify_peer}]), 
+...
+```
+We also plan to update the http client (`httpc`) to use this soon.
+
+# A new fast Pseudo Random Generator
+
+A new custom designed Pseudo Random Generator `rand:mwc59`
+has been implemented.  It is probably the fastest possible
+generator with good quality that can be written in Erlang.
+To do this it barely avoids bignums, allocating heap data,
+and uses only a minimal number of fast operations.
+
```

Comment:
Suggested addition:

Under the "right" circumstances: A number that takes 60 ns to generate
with the default generator can be generated in 4 ns with `mwc59`.

Reply:
Added this

Reply:
Added